src/cdk/explorer/detectCdkProjects.ts

```diff
@@ -0,0 +1,98 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as path from 'path'
+import * as vscode from 'vscode'
+import { access, stat } from '../../shared/filesystem'
+import { readDirAsString } from '../../shared/filesystemUtilities'
+
+export interface CdkProject {
+    workspaceFolder: vscode.WorkspaceFolder
+    cdkJsonPath: string
+    treePath: string
+}
+
+export async function detectCdkProjects(
+    workspaceFolders: vscode.WorkspaceFolder[] | undefined = vscode.workspace.workspaceFolders
+): Promise<CdkProject[]> {
+    if (!workspaceFolders) {
+        return []
+    }
+
+    return (await Promise.all(workspaceFolders.map(detectCdkProjectsFromWorkspaceFolder))).reduce(
+        (accumulator: CdkProject[], current: CdkProject[]) => {
+            accumulator.push(...current)
+
+            return accumulator
+        },
+        []
+    )
+}
+
+async function detectCdkProjectsFromWorkspaceFolder(workspaceFolder: vscode.WorkspaceFolder): Promise<CdkProject[]> {
+    const result = []
+
+    for await (const treeJson of detectLocalCdkProjects({ workspaceUris: [workspaceFolder.uri] })) {
+        const cdkJsonPath = path.join(treeJson.fsPath, '../', 'cdk.json')
+        const project = { workspaceFolder: workspaceFolder, cdkJsonPath: cdkJsonPath, treePath: treeJson.fsPath }
+        result.push(project)
+    }
+
+    return result
+}
+
+export interface DetectCdkProjectsContext {
+    access: typeof access
+    readDir: typeof readDirAsString
+    stat: typeof stat
+}
+
+class DefaultDetectCdkProjectsContext implements DetectCdkProjectsContext {
+    public readonly access = access
+    public readonly readDir = readDirAsString
+    public readonly stat = stat
+}
```

Comment:
I don't understand the point of these. These wrap general utility functions; why not just use them directly?

Reply:
I reused the pattern used in detecting sam local lambdas. I believe these exist for testing purposes and don't provide much utility otherwise. I think using something like [mockfs](http://mockfs.github.io/) is better but went with consistency of what was there.

Reply:
removed them